File: 2690-house-robber-iv/2690-house-robber-iv.java
class Solution {
    public boolean isPossible(int[] nums,int k,int cap){
        int cnt = 0,i = 0;
        while(i < nums.length){
            if(nums[i] <= cap){
                cnt++;
                i++;
            }
            i++;
        }
        if(cnt >= k)
            return true;
        return false;
    }
    public int minCapability(int[] nums, int k) {
        int l = Integer.MAX_VALUE,r = 0;
        for(int i:nums){
            r = Math.max(r,i);
            l = Math.min(l,i);
        }
        while(l <= r){
            int mid = (l + r) / 2;
            if(isPossible(nums,k,mid)){
                r = mid - 1;
            }else{
                l = mid + 1;
            }
        }
        return l;
    }
}

File: 2690-house-robber-iv/2690-house-robber-iv.py
class Solution:
    def isPossible(self, nums, k, cap):
        cnt = 0
        i = 0
        while i < len(nums):
            if nums[i] <= cap:
                cnt += 1  # Rob this house
                i += 1  # Skip the next house
            i += 1  # Move to next house
        return cnt >= k

    def minCapability(self, nums: List[int], k: int) -> int:
        l, r = min(nums), max(nums)

        while l <= r:
            mid = (l + r) // 2
            if self.isPossible(nums, k, mid):
                r = mid - 1
            else:
                l = mid + 1
        return l